reconcile/glitchtip_project_alerts/integration.py

```diff
@@ -0,0 +1,226 @@
+                        glitchtip_client.create_project_alert(
+                            organization_slug=org.slug,
+                            project_slug=desired_project.slug,
+                            alert=alert_to_add,
+                        )
+
+                for alert_to_remove in diff_result.delete.values():
+                    if not alert_to_remove.pk:
+                        # this can't happend - just make mypy happy
+                        continue
+                    logging.info(
+                        [
+                            "delete_project_alert",
+                            f"{org.name}/{desired_project.slug}/{alert_to_remove.name}",
+                        ]
+                    )
+                    if not dry_run:
+                        glitchtip_client.delete_project_alert(
+                            organization_slug=org.slug,
+                            project_slug=desired_project.slug,
+                            alert_pk=alert_to_remove.pk,
+                        )
+
+                for diff_pair in diff_result.change.values():
+                    alert_to_update = diff_pair.desired
+                    alert_to_update.pk = diff_pair.current.pk
+                    logging.info(
+                        [
+                            "update_project_alert",
+                            f"{org.name}/{desired_project.slug}/{alert_to_update.name}",
+                        ]
+                    )
+                    if not dry_run:
+                        glitchtip_client.update_project_alert(
+                            organization_slug=org.slug,
+                            project_slug=desired_project.slug,
+                            alert=alert_to_update,
+                        )
+
+    def run(self, dry_run: bool) -> None:
+        gqlapi = gql.get_api()
+        read_timeout, max_retries, _ = get_glitchtip_settings()
+        # data
+        glitchtip_instances = glitchtip_instance_query(
+            query_func=gqlapi.query
+        ).instances
+        glitchtip_projects = self.get_projects(query_func=gqlapi.query)
+
+        for glitchtip_instance in glitchtip_instances:
+            if self.params.instance and glitchtip_instance.name != self.params.instance:
+                continue
+
+            glitchtip_client = GlitchtipClient(
+                host=glitchtip_instance.console_url,
+                token=self.secret_reader.read_secret(
+                    glitchtip_instance.automation_token
+                ),
+                read_timeout=read_timeout,
+                max_retries=max_retries,
+            )
+            current_state = self.fetch_current_state(glitchtip_client=glitchtip_client)
+            desired_state = self.fetch_desired_state(
+                glitchtip_projects=[
+                    p
+                    for p in glitchtip_projects
+                    if p.organization.instance.name == glitchtip_instance.name
+                ]
```

Comment:
To avoid loop `glitchtip_projects` multiple times inside `glitchtip_instances` loop, can pre compute a `glitchtip_projects_by_instance`

```python
        glitchtip_projects_by_instance = defaultdict(list)
        for glitchtip_project in glitchtip_projects:
            glitchtip_projects_by_instance[
                glitchtip_project.organization.instance.name
            ].append(glitchtip_project)
```

then do a directly lookup

```python
            desired_state = self.fetch_desired_state(
                glitchtip_projects=glitchtip_projects_by_instance[
                    glitchtip_instance.name
                ]
            )
```

Reply:
Good idea. Applied in [7960685](https://github.com/app-sre/qontract-reconcile/pull/3763/commits/79606853a89ec881b489383deefc775bbf741e72)